constraints/miden-vm/decoder.air

```diff
@@ -0,0 +1,594 @@
+
+    # Enforce that when SPAN or RESPAN operations is executed, one of the batch flags must be set 
+    # to 1.
+    enf (f_span(op_bits) + get_f_respan(op_bits, extra)) - 
+        (f_g1 + f_g2 + f_g4 + get_f_g8(op_batch_flags)) = 0
+
+    # Enforce that when we have at most 4 groups in a batch, registers h[4], ..., h[7] should be 
+    # set to 0's.
+    enf (f_g1 + f_g2 + f_g4) * hasher[i] = 0 for i in 4..8
+
+    # Enforce that When we have at most 2 groups in a batch, registers h[2] and h[3] should also be
+    # set to 0's.
+    enf (f_g1 + f_g2) * hasher[i] = 0 for i in 2..4
+
+    # Enforce that when we have at most 1 groups in a batch, register h[1] should also be set to 0.
+    enf f_g1 * hasher[1] = 0
+
+
+# Enforce that all operation groups in a given batch are consumed before a new batch is started 
+# (i.e., via a RESPAN operation) or the execution of a span block is complete (i.e., via an END 
+# operation).
+#
+# Constraint degree: 9
+ev op_group_table_constraints([addr, op_bits[7], hasher[8], is_span, group_count, op_index, op_batch_flags[3], s0, extra], [p[4]]):
+    # Get value of the f_push flag
+    let f_push = get_f_push(op_bits, extra)
+
+    # Values of the rows to be added to the op group table when a SPAN or a RESPAN operation is 
+    # executed.
+    let v_i = $alpha[0] + $alpha[1] * addr' + $alpha[2] * (group_count - i) + $alpha[3] * hasher[i]
+
+    # Row value for group in hasher[1] to be added to the op group table when a SPAN or a RESPAN 
+    # operation is executed.
+    let v_1 = $alpha[0] + $alpha[1] * addr' + $alpha[2] * (group_count - 1) + $alpha[3] * hasher[1]
+
+    let prod_v_3 = prod([$alpha[0] + 
+                         $alpha[1] * addr' + 
+                         $alpha[2] * (group_count - i) + 
+                         $alpha[3] * hasher[i] for i in 1..4])
+
+    let prod_v_7 = prod([$alpha[0] + 
+                         $alpha[1] * addr' + 
+                         $alpha[2] * (group_count - i) + 
+                         $alpha[3] * hasher[i] for i in 1..8])
+
+    # The value of the row to be removed from the op group table.
+    let u = $alpha[0] + $alpha[1] * addr + $alpha[2] * group_count + $alpha[3] * 
+        ((hasher[0]' * 2^7 + op_index') * (1 - f_push + s0' * f_push)) = 0
+    
+    # A flag which is set to 1 when a group needs to be removed from the op group table.
+    let f_dg = is_span * (group_count' - group_count)
+
+    # Enforce the constraint for updating op group table. The constraint specifies that when SPAN 
+    # or RESPAN operations are executed, we add between 1 and 7 groups to the op group table, and 
+    # when group count is decremented inside a span block, we remove a group from the op group 
+    # table.
+    enf p[3]' * (f_dg * u + 1 - f_dg) = 
+        p[3] * (get_f_g2(op_batch_flags) * v_1 + get_f_g4(op_batch_flags) * 
+        prod_v_3 + get_f_g8(op_batch_flags) * 
+        prod_v_7 - 1 + (f_span(op_bits) + get_f_respan(op_bits, extra)))
```

Comment:
nit: I would do these line breaks at the `+` characters instead, to keep each set of values together